File: ios/MenCare/Core/Data/Network/NetworkCrmRepository.swift
import Foundation

@MainActor
final class NetworkCrmRepository: CrmRepository {

    /// La rubrica e' paginata a cursore; le schermate ne mostrano una lista
    /// piatta, quindi si chiede una pagina generosa e basta.
    private let pageSize = 100

    private let client: ApiClient

    init(client: ApiClient) {
        self.client = client
    }

    func clients(query: String, segment: ClientSegment) async throws -> [ClientRecord] {
        try await apiThrowing {
            let trimmed = query.trimmingCharacters(in: .whitespaces)
            let page = try await client.send(
                CrmEndpoint.clients(
                    query: trimmed.isEmpty ? nil : trimmed, segment: segment, limit: pageSize, cursor: nil
                ),
                as: ClientsPageDto.self
            )
            return page.clients.map { $0.toDomain() }
        }
    }

    func clientDetail(_ id: String) async throws -> ClientDetail {
        try await apiThrowing {
            let dto = try await client.send(CrmEndpoint.client(id), as: ClientDetailDto.self)
            var record = dto.client.toDomain()
            record.favoriteOperatorId = dto.insights?.favoriteOperatorId
            record.averageDaysBetweenVisits = dto.insights?.averageDaysBetweenVisits
            return ClientDetail(
                client: record,
                history: dto.appointments
                    .map { $0.toDomain(clientId: dto.client.id) }
                    .sorted { $0.start > $1.start }
            )
        }
    }

    func createClient(firstName: String, lastName: String, phone: String) async -> AppResult<ClientRecord> {
        await apiResult {
            try await client.send(
                CrmEndpoint.createClient(
                    CreateClientRequestDto(firstName: firstName, lastName: lastName, phone: phone, email: nil)
                ),
                as: ClientDto.self
            ).toDomain()
        }
    }
}

File: android/core/data/src/main/kotlin/com/devora/mencare/core/data/network/NetworkCrmRepository.kt
package com.devora.mencare.core.data.network

import com.devora.mencare.core.common.AppResult
import com.devora.mencare.core.common.DispatcherProvider
import com.devora.mencare.core.common.onSuccess
import com.devora.mencare.core.data.repository.CrmRepository
import com.devora.mencare.core.model.ClientRecord
import com.devora.mencare.core.model.ClientSegment
import com.devora.mencare.core.network.api.CrmApi
import com.devora.mencare.core.network.apiCall
import com.devora.mencare.core.network.dto.CreateClientBody
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.withContext
import javax.inject.Inject
import javax.inject.Singleton

/** Quanti clienti per pagina, e quante pagine al massimo si rincorrono. */
private const val PAGE_SIZE = 100
private const val MAX_PAGES = 10

/**
 * Rubrica clienti dal backend.
 *
 * Ricerca e segmenti li applica il server: "inattivi da 60 giorni" o "top
 * spesa" dipendono dallo storico intero, che l'app non ha e non deve avere.
 * La paginazione è a cursore: qui si scorre fino a [MAX_PAGES] perché le
 * schermate mostrano una lista, non un archivio — oltre, si cerca.
 */
@Singleton
class NetworkCrmRepository @Inject constructor(
    private val api: CrmApi,
    private val sync: DataSync,
    private val dispatchers: DispatcherProvider,
) : CrmRepository {

    override val clients: Flow<List<ClientRecord>> =
        sync.reloading(SyncKeys.CLIENTS, emptyList<ClientRecord>()) { fetchAll(null, ClientSegment.TUTTI) }

    override fun client(id: String): Flow<ClientRecord?> =
        sync.reloading<ClientRecord?>(SyncKeys.CLIENTS, null) {
            // Le abitudini arrivano solo con la scheda: si copiano sul record.
            val detail = api.client(id)
            detail.client.toModel().copy(
                favoriteOperatorId = detail.insights?.favoriteOperatorId,
                averageDaysBetweenVisits = detail.insights?.averageDaysBetweenVisits,
            )
        }

    override fun search(query: String, segment: ClientSegment): Flow<List<ClientRecord>> =
        sync.reloading(SyncKeys.CLIENTS, emptyList<ClientRecord>()) {
            fetchAll(query.trim().ifBlank { null }, segment)
        }

    private suspend fun fetchAll(query: String?, segment: ClientSegment): List<ClientRecord> {
        val found = mutableListOf<ClientRecord>()
        var cursor: String? = null
        repeat(MAX_PAGES) {
            val page = api.clients(query, segment.name, PAGE_SIZE, cursor)
            found += page.clients.map { it.toModel() }
            cursor = page.nextCursor ?: return found
        }
        return found
    }

    override suspend fun createClient(
        firstName: String,
        lastName: String,
        phone: String,
    ): AppResult<ClientRecord> = withContext(dispatchers.io) {
        apiCall { api.createClient(CreateClientBody(firstName, lastName, phone)).toModel() }
            .onSuccess { sync.invalidate() }
    }
}
